crates/biome_js_analyze/src/semantic_analyzers/nursery/no_misleading_character_class.rs

```diff
@@ -0,0 +1,646 @@
+    /// ```
+    ///
+    /// ## Valid
+    ///
+    /// ```js
+    /// /^[abc]$/;
+    /// /^[👍]$/u;
+    /// /^[\q{👶🏻}]$/v;
+    /// ```
+    ///
+
+    pub(crate) NoMisleadingCharacterClass {
+        version: "next",
+        name: "noMisleadingCharacterClass",
+        recommended: false,
+        fix_kind: FixKind::Safe,
+    }
+}
+
+declare_node_union! {
+  pub(crate) AnyRegexExpression = JsNewExpression | JsCallExpression | JsRegexLiteralExpression
+}
+
+pub enum Message {
+    SurrogatePairWithoutUFlag,
+    EmojiModifier,
+    RegionalIndicatorSymbol,
+    CombiningClassOrVs16,
+    JoinedCharSequence,
+}
+
+impl Message {
+    fn as_str(&self) -> &str {
+        match self {
+            Self::CombiningClassOrVs16 => "Unexpected combined character in character class.",
+            Self::SurrogatePairWithoutUFlag => {
+                "Unexpected surrogate pair in character class. Use 'u' flag."
+            }
+            Self::EmojiModifier => "Unexpected modified Emoji in character class. ",
+            Self::RegionalIndicatorSymbol => {
+                "Regional indicator symbol characters should not be used in character class."
+            }
+            Self::JoinedCharSequence => "Unexpected joined character sequence in character class.",
+        }
+    }
+}
+
+pub struct RuleState {
+    range: TextRange,
+    message: Message,
+}
+
+impl Rule for NoMisleadingCharacterClass {
+    type Query = Semantic<AnyRegexExpression>;
+    type State = RuleState;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let regex = ctx.query();
+
+        match regex {
+            AnyRegexExpression::JsRegexLiteralExpression(expr) => {
+                let Ok((pattern, flags)) = expr.decompose() else {
+                    return None;
+                };
+
+                if flags.text().contains('v') {
+                    return None;
+                }
+                let regex_pattern = replace_escaped_unicode(pattern.text());
+                let has_u_flag = flags.text().contains('u');
+                let range = expr.syntax().text_range();
+                return diagnostic_regex_pattern(&regex_pattern, has_u_flag, range);
+            }
+
+            AnyRegexExpression::JsNewExpression(expr) => {
+                let is_reg_exp = match expr.callee().ok()? {
+                    AnyJsExpression::JsIdentifierExpression(callee) => {
+                        callee.name().ok()?.has_name("RegExp")
+                    }
+                    AnyJsExpression::JsStaticMemberExpression(callee) => {
+                        let is_global_this = match callee.object().ok()? {
+                            AnyJsExpression::JsIdentifierExpression(e) => {
+                                e.name().ok()?.has_name("globalThis")
+                            }
+                            _ => false,
+                        };
```

Comment:
We have an utility called `global_identifier` for these cases, and you use this one. That's needed because it handles recursive cases of global identifiers, like this:

```js
globalThis.globalThis.globalThis.RegExp
```